src/clients/arclib/arcresub.1.in
.\" -*- nroff -*-
.TH ARCRESUB 1 "@DATE@" "NorduGrid ARC @VERSION@" "NorduGrid Users Manual"

.SH NAME

arcresub \- ARC Resubmission

.SH DESCRIPTION

The
.B arcresub
command is used for resubmitting jobs to grid enabled computing resources.

.SH SYNOPSIS

.B arcresub [options] [jobid ...]

.SH OPTIONS

.IP "\fB-a\fR, \fB--all\fR"
all jobs
.IP "\fB-j\fR, \fB--joblist\fR=\fIfilename\fR"
the file storing information about active jobs (default ~/.arc/jobs.xml)
.IP "\fB-i\fR, \fB--jobids-from-file\fR=\fIfilename\fR"
a file containing a list of job IDs
.IP "\fB-o\fR, \fB--jobids-to-file\fR=\fIfilename\fR"
the IDs of the submitted jobs will be appended to this file
.IP "\fB-c\fR, \fB--cluster\fR=\fIname\fR"
select one or more computing elements: \fIname\fR can be an alias for a single CE, a group of CEs or a URL
.IP "\fB-q\fR, \fB--qluster\fR=\fIname\fR"
select one or more computing elements for the new jobs: \fIname\fR can be an alias for a single CE, a group of CEs or a URL
.IP "\fB-g\fR, \fB--index\fR=\fIname\fR"
select one or more registries: \fIname\fR can be an alias for a single registry, a group of registries or a URL
.IP "\fB-r\fR, \fB--rejectmanagement\fR=\fIURL\fR"
skip jobs which are on a computing element with a given URL
.IP "\fB-R\fR, \fB--rejectdiscovery\fR=\fIURL\fR"
skip the service with the given URL during service discovery
.IP "\fB-S\fR, \fB--submissioninterface\fR=\fIInterfaceName\fR"
only use this interface for submitting (e.g. org.nordugrid.gridftpjob, org.ogf.glue.emies.activitycreation, org.ogf.bes)
.IP "\fB-I\fR, \fB--infointerface\fR=\fIInterfaceName\fR"
the computing element specified by URL at the command line should be
queried using this information interface (possible options:
org.nordugrid.ldapng, org.nordugrid.ldapglue2,
org.nordugrid.wsrfglue2, org.ogf.glue.emies.resourceinfo)
.IP "\fB-k\fR, \fB--keep\fR"
keep the files on the server (do not clean)
.IP "\fB-m\fR, \fB--same\fR"
resubmit to the same cluster
.IP "\fB-M\fR, \fB--not-same\fR"
do not resubmit to the same cluster
.IP "\fB-s\fR, \fB--status\fR=\fIstatusstr\fR"
only select jobs whose status is statusstr
.IP "\fB-P\fR, \fB--listplugins\fR"
list the available plugins
.IP "\fB-t\fR, \fB--timeout\fR=\fIseconds\fR"
timeout in seconds (default 20)
.IP "\fB-z\fR, \fB--conffile\fR=\fIfilename\fR"
configuration file (default ~/.arc/client.conf)
.IP "\fB-d\fR, \fB--debug\fR=\fIdebuglevel\fR"
FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG
.IP "\fB-b\fR, \fB--broker\fR=\fIbroker\fR"
selected broker: Random (default), FastestQueue or custom
.IP "\fB-v\fR, \fB--version\fR"
print version information
.IP "\fB-?\fR, \fB--help\fR"
print help
.LP

.SH ARGUMENTS

.IP "\fBjobid\fR ..."
job ID or job name of jobs to be resubmitted
.LP

.SH EXTENDED DESCRIPTION

The
.B arcresub
command is used for resubmitting jobs on grid enabled computing
resources.
The job can be referred to either by the job ID that was returned by
.BR arcsub (1)
at submission time or by its jobname if the job description that was
submitted contained a jobname attribute.

More than one job ID and/or jobname can be given. If several jobs were
submitted with the same jobname all those jobs will be resubmitted.

If the job description of a job to be resubmitted, contained any local
input files, checksums of these was calculated and stored in the
job list, and those will be used to check whether the files has changed.
If local input files has changed the job will not be resubmitted.

In case the job description is not found in the job list, it will
be retrieved from the cluster holding the job. This description
however might differ from the one originally used to submit the job,
since both the submission client and the cluster can have made
modifications to the job description.

Upon resubmision the job will receive a new job ID, and the old
job ID will be stored in the local job list file, enabling future
back tracing of the resubmitted job.

The name of the local job list file can used specified by the
.B --joblist
option. By specifying the
.B --all
option, all active jobs appearing the in job list file will be
resubmitted.

The
.B --cluster
option can be used to select or reject jobs at specific clusters to be
resubmitted. See
.BR arcsub (1)
for a discussion of the format of arguments to this option.
The
.B --qluster
option can be used to force the jobs to be resubmitted to a particular
cluster, or to reject resubmission to a particular cluster. Again see
.BR arcsub (1)
for a discussion of the format of arguments to this option.
The
.B --status
option can be used to select jobs in a specific state. These options
can be repeated several times.
Which servers to query can be specified by giving the
.B --index
option to the command. See
.BR arcsub (1)
for a discussion of the format of arguments to this option.

The default behaviour of
.BR arcresub
is to resubmit to any cluster. This behaviour can be changed by
specifying the \fB--same\fR or \fB--not-same\fR options. The
former will resubmit a job to the same cluster as the job resides,
and thus the \fB--qluster\fR and \fB--index\fR options is ignored.
The latter will resubmit a job to any cluster except to the
cluster it resides. Note the \fB--same\fR and \fB--not-same\fR
options cannot be specified together.

If the old job was successfully killed the job will be removed from the
remote cluster unless the \fB--keep\fR option was specified.

.SH FILES

.TP
.B ~/.arc/client.conf
Some options can be given default values by specifying them in the
ARC client configuration file. By using the
.B --conffile
option a different configuration file can be used than the default.

.TP
.B ~/.arc/jobs.xml
This a local list of the user's active jobs. When a job is
successfully submitted it is added to this list and when it is removed
from the remote cluster it is removed from this list. This list is
used as the list of all active jobs when the user specifies the

.SH ENVIRONMENT VARIABLES

.TP
.B X509_USER_PROXY
The location of the user's grid proxy file. Shouldn't be set unless
the proxy is in a non-standard location.

.TP
.B ARC_LOCATION
The location where ARC is installed can be specified by this variable. If not
specified the install location will be determined from the path to the
command being executed, and if this fails a WARNING will be given stating the
location which will be used.

.TP
.B ARC_PLUGIN_PATH
The location of ARC plugins can be specified by this variable. Multiple
locations can be specified by separating them by : (; in Windows). The
default location is \fB$ARC_LOCATION\fR/lib/arc (\\ in Windows).

.SH COPYRIGHT

APACHE LICENSE Version 2.0

.SH AUTHOR

Mattias Ellert <mattias.ellert@fysast.uu.se>

.SH SEE ALSO

.BR arccat (1),
.BR arcclean (1),
.BR arccp (1),
.BR arcget (1),
.BR arcinfo (1),
.BR arckill (1),
.BR arcls (1),
.BR arcmigrate (1),
.BR arcmkdir (1),
.BR arcproxy (1),
.BR arcrenew (1),
.BR arcresume (1),
.BR arcrm (1),
.BR arcstat (1),
.BR arcsub (1),
.BR arcsync (1),
.BR arctest (1)



NEWS
no news good news :)


src/clients/arclib/arcresub.cpp
// -*- indent-tabs-mode: nil -*-

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <list>
#include <string>
#include <sys/types.h>
#include <sys/stat.h>

#include <arc/ArcLocation.h>
#include <arc/IString.h>
#include <arc/Logger.h>
#include <arc/URL.h>
#include <arc/UserConfig.h>
#include <arc/compute/Job.h>
#include <arc/compute/JobSupervisor.h>

#include "utils.h"

int RUNMAIN(arcresub)(int argc, char **argv) {

  setlocale(LC_ALL, "");

  Arc::Logger logger(Arc::Logger::getRootLogger(), "arcresub");
  Arc::LogStream logcerr(std::cerr);
  logcerr.setFormat(Arc::ShortFormat);
  Arc::Logger::getRootLogger().addDestination(logcerr);
  Arc::Logger::getRootLogger().setThreshold(Arc::WARNING);

  Arc::ArcLocation::Init(argv[0]);

  ClientOptions opt(ClientOptions::CO_RESUB, istring("[job ...]"));

  std::list<std::string> jobIDsOrNames = opt.Parse(argc, argv);

  if (opt.showversion) {
    std::cout << Arc::IString("%s version %s", "arcresub", VERSION)
              << std::endl;
    return 0;
  }

  // If debug is specified as argument, it should be set before loading the configuration.
  if (!opt.debug.empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(opt.debug));

  Arc::UserConfig usercfg(opt.conffile, opt.joblist);
  if (!usercfg) {
    logger.msg(Arc::ERROR, "Failed configuration initialization");
    return 1;
  }

  if (opt.show_plugins) {
    std::list<std::string> types;
    types.push_back("HED:SubmitterPlugin");
    types.push_back("HED:ServiceEndpointRetrieverPlugin");
    types.push_back("HED:TargetInformationRetrieverPlugin");
    types.push_back("HED:BrokerPlugin");
    showplugins("arcresub", types, logger, usercfg.Broker().first);
    return 0;
  }

  if (!checkproxy(usercfg)) {
    return 1;
  }

  if (opt.debug.empty() && !usercfg.Verbosity().empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(usercfg.Verbosity()));

  if (opt.same && opt.notsame) {
    logger.msg(Arc::ERROR, "--same and --not-same cannot be specified together.");
    return 1;
  }

  for (std::list<std::string>::const_iterator it = opt.jobidinfiles.begin(); it != opt.jobidinfiles.end(); it++) {
    if (!Arc::Job::ReadJobIDsFromFile(*it, jobIDsOrNames)) {
      logger.msg(Arc::WARNING, "Cannot read specified jobid file: %s", *it);
    }
  }

  if (opt.timeout > 0)
    usercfg.Timeout(opt.timeout);

  if (!opt.broker.empty())
    usercfg.Broker(opt.broker);

  if ((!opt.joblist.empty() || !opt.status.empty()) && jobIDsOrNames.empty() && opt.clusters.empty())
    opt.all = true;

  if (jobIDsOrNames.empty() && opt.clusters.empty() && !opt.all) {
    logger.msg(Arc::ERROR, "No jobs given");
    return 1;
  }

  std::list<std::string> selectedURLs = getSelectedURLsFromUserConfigAndCommandLine(usercfg, opt.clusters);
  std::list<std::string> rejectManagementURLs = getRejectManagementURLsFromUserConfigAndCommandLine(usercfg, opt.rejectmanagement);

  std::list<Arc::Job> jobs;
  if (!Arc::Job::ReadJobsFromFile(usercfg.JobListFile(), jobs, jobIDsOrNames, opt.all, selectedURLs, rejectManagementURLs)) {
    logger.msg(Arc::ERROR, "Unable to read job information from file (%s)", usercfg.JobListFile());
    return 1;
  }

  for (std::list<std::string>::const_iterator itJIDOrName = jobIDsOrNames.begin();
       itJIDOrName != jobIDsOrNames.end(); ++itJIDOrName) {
    std::cout << Arc::IString("Warning: Job not found in job list: %s", *itJIDOrName) << std::endl;
  }


  Arc::JobSupervisor jobmaster(usercfg, jobs);
  jobmaster.Update();
  jobmaster.SelectValid();
  if (!opt.status.empty()) {
    jobmaster.SelectByStatus(opt.status);
  }

  if (jobmaster.GetSelectedJobs().empty()) {
    std::cout << Arc::IString("No jobs") << std::endl;
    return 1;
  }

  std::list<Arc::Endpoint> services = getServicesFromUserConfigAndCommandLine(usercfg, opt.qlusters, opt.indexurls, opt.requestedSubmissionInterfaceName, opt.infointerface);
  std::list<std::string> rejectDiscoveryURLs = getRejectDiscoveryURLsFromUserConfigAndCommandLine(usercfg, opt.rejectdiscovery);

  std::list<Arc::Job> resubmittedJobs;
  // same + 2*notsame in {0,1,2}. same and notsame cannot both be true, see above.
  int retval = (int)!jobmaster.Resubmit((int)opt.same + 2*(int)opt.notsame, services, resubmittedJobs, rejectDiscoveryURLs);
  if (retval == 0 && resubmittedJobs.empty()) {
    std::cout << Arc::IString("No jobs to resubmit with the specified status") << std::endl;
    return 0;
  }

  for (std::list<Arc::Job>::const_iterator it = resubmittedJobs.begin();
       it != resubmittedJobs.end(); ++it) {
    std::cout << Arc::IString("Job submitted with jobid: %s", it->JobID) << std::endl;
  }

  if (!resubmittedJobs.empty() && !Arc::Job::WriteJobsToFile(usercfg.JobListFile(), resubmittedJobs)) {
    std::cout << Arc::IString("Warning: Failed to lock job list file %s", usercfg.JobListFile()) << std::endl;
    std::cout << Arc::IString("         To recover missing jobs, run arcsync") << std::endl;
    retval = 1;
  }

  if (!opt.jobidoutfile.empty() && !Arc::Job::WriteJobIDsToFile(resubmittedJobs, opt.jobidoutfile)) {
    logger.msg(Arc::WARNING, "Cannot write jobids to file (%s)", opt.jobidoutfile);
    retval = 1;
  }

  std::list<std::string> notresubmitted = jobmaster.GetIDsNotProcessed();

  if (!jobmaster.Cancel()) {
    retval = 1;
  }
  for (std::list<std::string>::const_iterator it = jobmaster.GetIDsNotProcessed().begin();
       it != jobmaster.GetIDsNotProcessed().end(); ++it) {
    logger.msg(Arc::WARNING, "Resubmission of job (%s) succeeded, but killing the job failed - it will still appear in the job list", *it);
  }

  if (!opt.keep) {
    if (!jobmaster.Clean()) {
      retval = 1;
    }
    for (std::list<std::string>::const_iterator it = jobmaster.GetIDsNotProcessed().begin();
         it != jobmaster.GetIDsNotProcessed().end(); ++it) {
      logger.msg(Arc::WARNING, "Resubmission of job (%s) succeeded, but cleaning the job failed - it will still appear in the job list", *it);
    }

    if (!Arc::Job::RemoveJobsFromFile(usercfg.JobListFile(), jobmaster.GetIDsProcessed())) {
      std::cout << Arc::IString("Warning: Failed to lock job list file %s", usercfg.JobListFile()) << std::endl;
      std::cout << Arc::IString("         Use arcclean to remove non-existing jobs") << std::endl;
      retval = 1;
    }
  }

  if ((resubmittedJobs.size() + notresubmitted.size()) > 1) {
    std::cout << std::endl << Arc::IString("Job resubmission summary:") << std::endl;
    std::cout << "-----------------------" << std::endl;
    std::cout << Arc::IString("%d of %d jobs were resubmitted", resubmittedJobs.size(), resubmittedJobs.size() + notresubmitted.size()) << std::endl;
    if (!notresubmitted.empty()) {
      std::cout << Arc::IString("The following %d were not resubmitted", notresubmitted.size()) << std::endl;
      for (std::list<std::string>::const_iterator it = notresubmitted.begin();
           it != notresubmitted.end(); ++it) {
        std::cout << *it << std::endl;
      }
    }
  }

  return retval;
}
